File: SyncScript.Core/CRDTDocument.cs
using System;
using System.Collections.Generic;
using System.Text;

namespace SyncScript.Core
{
    public class CRDTDocument
    {
        private readonly LinkedList<CRDTNode> _nodes = new();
        private readonly Dictionary<Guid, LinkedListNode<CRDTNode>> _index = new();
        public void Insert(CRDTNode node)
        {
            if (_index.ContainsKey(node.Id)) return;
            if(node.ParentId == null)
            {
                var listNode = _nodes.AddFirst(node);
                _index[node.Id] = listNode;
            }
            else
            {
                if(!_index.TryGetValue(node.ParentId.Value,out var parentListNode))
                {
                    throw new InvalidOperationException("Parent node not found");
                }
                var postion = parentListNode;
                while(postion.Next != null && postion.Next.Value.ParentId  == node.ParentId
                    && postion.Next.Value.Id.CompareTo(node.Id) < 0)
                {
                    postion = postion.Next;
                }

                var listNode = _nodes.AddAfter(parentListNode, node);
                    _index[node.Id] = listNode;
                
            }
            
        }
        public void Delete(Guid nodeId)
        {
            if(_index.TryGetValue(nodeId,out var listNode))
            {
                listNode.Value.Tombstoned = true;
            }
        }

        public void Apply(CRDTOperation operation)
        {
            switch (operation.Type)
            {
                case OperationType.Insert:
                    Insert(operation.Node);
                    break;
                    Delete(operation.Node.Id);
                    break;
            }
        }
        public string GetText()
        {
            var result = new System.Text.StringBuilder();
            foreach(var node in _nodes)
            {
                if (!node.Tombstoned)
                    result.Append(node.Value);
            }
            return result.ToString();
        }
    }
}

File: syncscript-client/src/crdt/CRDTDocument.ts
import type{ CRDTNode, CRDTOperation } from '../types/crdt';

export class CRDTDocument {
  private nodes: CRDTNode[] = []

  insert(node: CRDTNode): void {
    if (this.nodes.find(n => n.id === node.id)) return

    if (node.parentId === null) {
      this.nodes.unshift(node)
    } else {
      const parentIndex = this.nodes.findIndex(n => n.id === node.parentId)
      if (parentIndex === -1) throw new Error('Parent node not found')

      let position = parentIndex
      while (
        position + 1 < this.nodes.length &&
        this.nodes[position + 1].parentId === node.parentId &&
        this.nodes[position + 1].id < node.id
      ) {
        position++
      }

      this.nodes.splice(position + 1, 0, node)
    }
  }

  delete(nodeId: string): void {
    const node = this.nodes.find(n => n.id === nodeId)
    if (node) node.tombstoned = true
  }

  apply(operation: CRDTOperation): void {
    if (operation.type === 'Insert') {
      this.insert(operation.node)
    } else {
      this.delete(operation.node.id)
    }
  }

  getNodes(): CRDTNode[] {
    return this.nodes.filter(n => !n.tombstoned)
  }
}